### PR TITLE
Fix MinHash + Add Scalable BF #28 + Optimize Partitionned BF

### PR DESCRIPTION
**Pull request by folkvir**
* Fix MinHash, comparison was wrong due to a unique lodash filter.
* Optimize the Partitionned Bloom Filter with a list of BitSet, import is still doable with the old structure to the new one.
  * Remove the length function
* Add Scalable Bloom Filter #28

A major bump is required.

### CODE REVIEW

**Comment by folkvir** [comment]
@Callidon Any idea to recover the length function from Partitionned Bloom Filter?
The PR is in draft until everything is ok, changelog, documentation, minor changes, etc...

File: README.md [hunk 1]
```diff
@@ -563,16 +600,17 @@ See [documentation online](https://callidon.github.io/bloom-filters/) or generat
...
 
 When submitting pull requests please follow the following guidance:
-* please, PRs on the develop branch. PRs on the master will be refused without comment.
-* add tests when possible in `./test/*-test.js` or by creating `*-test.js` files.
-* functions, methods, variables and types must be documented using typedoc annotations
-* run `yarn lint` to see linting errors
-* run `yarn fix` to fix errors that can be auto-fixed by gts.
-* run `yarn test` (build, lint and run the mocha tests suite with code coverage)
+
+- please, PRs on the develop branch. PRs on the master will be refused without comment.
```
Comment (Callidon) [inline README.md:608]: It's a bit of nitpicking, but I would rephrase to "Please open pull requests on the `develop` branch. Direct contributions to the `master` branch will be refused without comments. Also, each sentence in the enumeration is missing a capital letter in its first word.

File: src/bloom/partitioned-bloom-filter.ts [hunk 2]
```diff
@@ -107,12 +108,28 @@ export default class PartitionedBloomFilter
...
   public _m: number
-  @Field()
-  public _filter: Array<Array<number>>
+  @Field(
+    (sets: BitSet[]) => sets.map(s => s.export()),
+    (array: Array<{size: number; content: string} | number[]>) =>
+      array.map((data: {size: number; content: string} | number[]) => {
+        if (Array.isArray(data)) {
+          // Create the bitset from new and old exported structure
+          // create a new BitSet from the specified array
```
Comment (Callidon) [inline src/bloom/partitioned-bloom-filter.ts:117]: I think this comment should go around line 126, as it is about the `else` case of the condition

File: src/bloom/scalable-bloom-filter.ts [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+'use strict'
```
Comment (Callidon) [inline src/bloom/scalable-bloom-filter.ts:25]: Since we switched to typescript, I'm not sure we need to redeclare the strict mode for each file (old habits die hard), and we can just toggle it as a global setting in the typescript compiler options.

File: src/bloom/scalable-bloom-filter.ts [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+ */
+@AutoExportable<ScalableBloomFilter>('ScalableBloomFilter', ['_seed'])
+export default class ScalableBloomFilter
+  extends BaseFilter
+  implements ClassicFilter<HashableInput>
+{
+  /**
+   * Static value, will power the size of the new set, by default we will follow a power of 2.
+   */
+  public static _s = 2
```
Comment (Callidon) [inline src/bloom/scalable-bloom-filter.ts:50]: I suppose you named it `_s` because the algorithm also uses this name. If so, it's fine, I prefer that the code stick as much as possible to the original algorithms. Otherwise, I think we should aim for a more explicit name.

File: src/bloom/scalable-bloom-filter.ts [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+   */
+  @Field()
+  public _ratio: number
+
+  /**
+   * Internal Partition Bloom Filters
+   */
+  @Field(
+    (filters: PartitionBloomFilter[]) =>
+      filters.map(filter => filter.saveAsJSON()), // eslint-disable-line @typescript-eslint/no-unsafe-return
```
Comment (Callidon) [inline src/bloom/scalable-bloom-filter.ts:75]: Another reminder that our export/import system barely works 😄 But it's not the subject here, it's fine for me.

File: src/bloom/scalable-bloom-filter.ts [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+  }
+
+  /**
+   * @override
+   * Set the seed for this structure. If you override the seed after adding data,
+   * all the filters will be updated and you may get wrong indexes for already indexed data
+   * due to the seed change. So only change it once before adding data.
+   * @param  seed the new seed that will be used in this structure
+   */
+  public set seed(seed: number) {
```
Comment (Callidon) [inline src/bloom/scalable-bloom-filter.ts:121]: Can't we move this logic in a super class or a mixin? It seems common to all structures. Same goes for the getter btw.

File: src/bloom/scalable-bloom-filter.ts [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+   * @returns
+   */
+  public has(element: HashableInput) {
+    let i = 0
+    let found = false
+    while (!found && i < this._filters.length) {
+      found = this._filters[i].has(element)
+      i++
+    }
+    return found
```
Comment (Callidon) [inline src/bloom/scalable-bloom-filter.ts:166]: I think the following is more compact and strictly equivalent
```suggestion
    return this._filters.some((filter) => filter.has(element))
```

File: src/bloom/scalable-bloom-filter.ts [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+      return false
+    }
+    // check the ratio
+    if (this._ratio !== filter._ratio) {
+      return false
+    }
+    // check the capacity
+    if (this.capacity() !== filter.capacity()) {
+      return false
+    }
```
Comment (Callidon) [inline src/bloom/scalable-bloom-filter.ts:202]: Why not merge all conditions in the same if, as they all leads to the same return value?
```suggestion
    // assert the seed, the ratio and the capacity are equals
    if (this.seed !== filter.seed || this._ratio !== filter._ratio || this.capacity() !== filter.capacity()) {
      return false
    }
```

File: src/sketch/min-hash.ts [hunk 9]
```diff
@@ -125,7 +122,16 @@ export class MinHash extends BaseFilter {
       const candidateSignatures = values.map((value: number) =>
         applyHashFunction(value, this._hashFunctions[i])
       )
-      this._signature[i] = Math.min(this._signature[i], ...candidateSignatures)
+      // get the minimum of the candidate Signatures
+      // dont supply too much parameters to Math.min or Math.max with risk of getting stack error
+      // so we compute an itrative minmium
```
Comment (Callidon) [inline src/sketch/min-hash.ts:127]: typos
```suggestion
      // so we compute an iterative minimum
```

**Review by Callidon — CHANGES_REQUESTED** [review]
I have some comments and suggested changes, but otherwise it looks great. It will be a nice additions to the library, and fixes some impactful bugs. Good job 👏 👍

File: src/bloom/scalable-bloom-filter.ts (hunk 3 above)
Comment (folkvir) [reply src/bloom/scalable-bloom-filter.ts:25]: Strict checking is already enable in the tsconfig and tests are ok without `'use strict'`. I will remove the line everywhere!

File: src/bloom/scalable-bloom-filter.ts (hunk 5 above)
Comment (folkvir) [reply src/bloom/scalable-bloom-filter.ts:75]: Yes it should auto export based on the type, could be:
1) object or primitive type => check the class
2) array/iterator => iterate and go to 1)

File: src/bloom/scalable-bloom-filter.ts (hunk 4 above)
Comment (folkvir) [reply src/bloom/scalable-bloom-filter.ts:50]: Yes, it is according to the paper the same variable. By default with this setting we will follow a power of 2. Changing `_s` to N will follow a power of N. There is no reason to change this settings for now, this is why the variable is _static but not readonly.

File: src/bloom/scalable-bloom-filter.ts (hunk 7 above)
Comment (folkvir) [reply src/bloom/scalable-bloom-filter.ts:166]: `.some` iterates only on a fixed array and do not iterate on inserted elements after the iteration begins while the `while` loop does allow it. For mono-threaded environment your suggestion is ok. +1

File: src/bloom/scalable-bloom-filter.ts (hunk 8 above)
Comment (folkvir) [reply src/bloom/scalable-bloom-filter.ts:202]: And yours is less expensive too.

File: src/bloom/scalable-bloom-filter.ts (hunk 6 above)
Comment (folkvir) [reply src/bloom/scalable-bloom-filter.ts:119]: Logic is already in the base filter class we already extend : https://github.com/Callidon/bloom-filters/blob/f580c3caa30bb74338aaae2110fcdb76e6f25a49/src/base-filter.ts#L59-L64

But sometimes, the seed has to be propagated and it is the only way to override it correctly for the Scalable BF

File: src/bloom/scalable-bloom-filter.ts (hunk 3 above)
Comment (folkvir) [reply src/bloom/scalable-bloom-filter.ts]: commit [9332523](https://github.com/Callidon/bloom-filters/pull/53/commits/93325239aacf3b3bbd613bb7babacba9a0418fcd)

File: README.md (hunk 1 above)
Comment (folkvir) [reply README.md]: commit [ef6c6f3](https://github.com/Callidon/bloom-filters/pull/53/commits/ef6c6f381c8dcee1cc6e3aa236ffb09969b18839)

File: src/bloom/partitioned-bloom-filter.ts (hunk 2 above)
Comment (folkvir) [reply src/bloom/partitioned-bloom-filter.ts]: documentation updated in [b7e8fa5](https://github.com/Callidon/bloom-filters/pull/53/commits/b7e8fa56b86b4008607281b8a5957ade1aa7e0a7)

**Review by Callidon — APPROVED** [review]
Thank for the changes, all is good for me, we can merge it alongside #54 for the next release
